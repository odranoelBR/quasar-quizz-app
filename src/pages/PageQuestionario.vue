<template>
  <q-page
    id="questionario"
    v-if="getCurrentQuestion"
  >
    <q-separator
      dark
      class="q-mb-md"
    />

    <form-questionario
      v-if="getCurrentQuestion.id"
      @reset="reset"
      @update="updateCurrent"
      @disable-analise="setDisableAnalise"
      @created="getImagem"
      @resposta-analisada="setRespostaAnalisada "
      :answer="answer"
      :current-question="getCurrentQuestion"
      :resposta-analisada="respostaAnalisada"
      :url-for-image="urlForImage"
    />

    <div class="row q-pa-xs q-gutter-sm justify-between text-right">
      <transition
        appear
        :duration="{ enter: 300, leave: 300 }"
        mode="out-in"
        enter-active-class="animated fadeIn"
        leave-active-class="animated fadeOut"
      >
        <div class="col-auto">
          <q-btn
            style="height: 41px"
            color="primary"
            v-if="!disableAnalise && algumaRespostaSelecionada"
            @click="analisar"
          >
            Analisar
          </q-btn>
        </div>
      </transition>

      <div class="col">
        <q-btn
          color="accent"
          style="height: 41px"
          @click="back"
          v-show="!ehPrimeiraQuestao"
        >
          <q-icon name="arrow_back" />
        </q-btn>
      </div>
      <div class="col-auto">
        <q-btn
          color="accent"
          style="height: 41px"
          @click="next"
          v-show="!ehUltimaQuestao"
        >
          <q-icon name="arrow_forward" />
        </q-btn>
      </div>
    </div>
    <q-dialog
      full-height
      full-width
      v-model="dialog"
    >
      <q-card class="column justify-between bg-dialog">
        <q-card-section class="no-padding">
          <div style="min-height: 250px">
            <transition
              :duration="5000"
              enter-active-class="animated slideInLeft"
              leave-active-class="animated slideOutLeft"
            >
              <img
                v-if="trophy"
                class="q-pt-lg q-pb-lg"
                src="statics/success.svg"
                alt=""
              >
            </transition>
          </div>

          <div class="row no-wrap items-center text-center">
            <div class="col text-h4 ellipsis text-bold">
              Parabéns!
            </div>
          </div>
        </q-card-section>

        <q-card-section class="col text-center bg-dialog q-pt-sm">
          <div class="text-h3 text-positive">
            {{ porcentagemAcertos }}%
            Acerto
          </div>
          <div class="text-h6 q-pt-md">
            Você completou as questões!
          </div>
          <div class="text-h6">
            Foi <span class="text-h5 text-accent">{{ respostasCorretas }}</span> acerto
            de    <span class="text-h5 text-accent">{{ configQuestionary.qtdQuestoes }}</span>
          </div>
        </q-card-section>

        <q-card-section class="column ">
          <q-btn
            class="answer-chip q-mb-md"
            color="secondary"
            @click="$router.push('historico')"
          >
            Histórico
          </q-btn>

          <q-btn
            class="answer-chip "
            color="accent"
            v-close-popup
            @click="$router.push(`modulo/${choosedQuestionary.id}`)"
          >
            Novo caderno
          </q-btn>
        </q-card-section>
      </q-card>
    </q-dialog>
  </q-page>

  <q-page v-else-if="!loading">
    <div-sem-questoes-cadastradas />
  </q-page>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import { mapFields } from 'vuex-map-fields'
import FormQuestionario from 'components/FormQuestionario'
import DivSemQuestoesCadastradas from 'components/DivSemQuestoesCadastradas'
// import Vivus from 'vivus'
export default {
  created () {
    this.getQuestions()
  },
  components: {
    FormQuestionario, DivSemQuestoesCadastradas
  },
  data: () => ({
    respostaCorreta: { icon: 'thumb_up', color: 'positive' },
    respostaErrada: { icon: 'thumb_down', color: 'negative' },
    disableAnalise: false,
    respostaAnalisada: false,
    dialog: false,
    trophy: false
  }),
  watch: {
    dialog () {
      if (this.dialog) {
        setTimeout(() => {
          this.trophy = true
        }, 2)
      }
    },
    cadernoEstaFinalizado () {
      if (this.cadernoEstaFinalizado) {
        this.dialog = true
      }
    }
  },
  computed: {
    ...mapGetters('questionario', ['getCurrentQuestion', 'ehUltimaQuestao', 'ehPrimeiraQuestao', 'cadernoEstaFinalizado']),
    ...mapFields(['loading']),
    ...mapFields('questionario', ['configQuestionary', 'answers', 'choosedQuestionary', 'urlForImage']),
    algumaRespostaSelecionada () {
      return this.getCurrentQuestion.respostas.some(resposta => resposta.selecionada)
    },
    answer () {
      return this.answers.find(answer => answer.idQuestao === this.getCurrentQuestion.id)
    },
    respostasCorretas () {
      return this.answers.filter(modulo => modulo.correta).length
    },
    porcentagemAcertos () {
      return (this.respostasCorretas / this.configQuestionary.qtdQuestoes * 100)
        .toString().substr(0, 5)
    }
  },
  methods: {
    ...mapActions('questionario', ['getQuestions', 'updateAnswer', 'nextQuestion', 'backQuestion', 'updateCurrentQuestionChoice', 'resetChoices', 'getImage']),
    setDisableAnalise (flag) { this.disableAnalise = flag },
    setRespostaAnalisada (flag) { this.respostaAnalisada = flag },
    reset (respostaIndex) {
      this.resetChoices(respostaIndex)
    },
    updateCurrent (respostaIndex) {
      this.updateCurrentQuestionChoice(respostaIndex)
    },
    back () {
      this.respostaAnalisada = false
      this.backQuestion()
    },
    next () {
      this.respostaAnalisada = false
      this.nextQuestion()
    },
    analisar () {
      this.respostaAnalisada = true
      this.disableAnalise = true
      let answer = this.getCurrentQuestion.respostas.find(resposta => resposta.selecionada)
      this.updateAnswer({ idQuestao: this.getCurrentQuestion.id, letra: answer.letra, modulo: this.getCurrentQuestion.modulo, correta: answer.correta })
      this.$q.notify({
        ...{ position: 'bottom-right', classes: 'notify-questionario', group: false }, ...(answer.correta ? this.respostaCorreta : this.respostaErrada)
      })
    },
    showBizu () {
      this.$q.notify({
        message: this.getCurrentQuestion.bizu,
        color: 'primary',
        position: 'top',
        timeout: '3000',
        avatar: 'statics/idea.svg'
      })
    },
    getImagem () {
      if (this.getCurrentQuestion.image) {
        this.getImage(this.getCurrentQuestion.id)
      }
    }

  }
}
</script>
<style scoped>
[v-cloak] {
  display: none;
}
</style>
