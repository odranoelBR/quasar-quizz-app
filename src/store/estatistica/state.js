export const state = {
  total: 113,
  atual: {
    'Escala de serviço': 12,
    'Férias': 10,
    'Moviment. Competências': 9,
    'Licenças': 9,
    'Atribuições de serviços': 7,
    'Situações de OM': 6,
    'Substituição de agentes': 6,
    'Insígnias': 6,
    'Afastamentos de serviço': 6,
    'Revista / Formatura': 5,
    'Incl. Excl. Desligamento': 4,
    'Adição': 4,
    'Arrolamento de bens': 3,
    'Dispensas de serviço': 3,
    'Galeria de retratos': 3,
    'Apresentação': 2,
    'Falecimento': 2,
    'Serviço em equipe': 2,
    'Serviço externo': 1,
    'Bandeira nacional': 1,
    'Estandarte e emblemas': 1
  }
}
