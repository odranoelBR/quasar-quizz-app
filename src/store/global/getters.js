export function getUsuario (state) {
  return state.usuario
}
export function getModulos (state) {
  return state.modulos
}
