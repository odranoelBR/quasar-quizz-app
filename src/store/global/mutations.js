export function setUsuario (state, payload) {
  state.usuario = payload
}
