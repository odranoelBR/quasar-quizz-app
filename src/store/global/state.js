export default {
  user: {
    errorSound: false,
    tutorial: false,
    rightHand: true
  },
  modulos: [],
  loading: false
}
