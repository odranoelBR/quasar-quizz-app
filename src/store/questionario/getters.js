export function getCurrentQuestion (state) {
  return state.questions[state.currentQuestionIndex]
}
export function getCurrentQuestionIndex (state) {
  return state.currentQuestionIndex
}
export function getChoosedQuestionary (state) {
  return state.choosedQuestionary
}
export function getConfigQuestionary (state) {
  return state.configQuestionary
}
