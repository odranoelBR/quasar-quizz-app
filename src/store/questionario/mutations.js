import types from './types'

export default {
  [types.SET_ANSWERS] (state, payload) {
    state.answers = payload
  },
  [types.SET_QUESTIONS] (state, payload) {
    state.questions = payload
  },
  [types.SET_CURRENT_QUESTION_INDEX] (state, payload) {
    state.currentQuestionIndex = payload
  },
  [types.SET_CONFIG_QUESTIONARY] (state, payload) {
    state.configQuestionary = payload
  },
  [types.UPDATE_CURRENT_ANSWER_CHOICE] (state, payload) {
    state.questions[state.currentQuestionIndex].respostas[payload].selecionada = true
  }

}
