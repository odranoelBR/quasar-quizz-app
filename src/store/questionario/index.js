import { firestoreAction } from 'vuexfire'
import { db } from 'boot/firebase'

import state from './state'
import * as getters from './getters'
import * as mutations from './mutations'
import * as actions from './actions'

export default {
  namespaced: true,
  getters,
  mutations,
  actions: {
    ...actions,
    bindAnswers: firestoreAction(({ bindFirestoreRef, rootState }) => {
      return bindFirestoreRef('answers', db.collection('respostas').where('idUsuario', '==', rootState.usuario.id))
    })
  },
  state
}
