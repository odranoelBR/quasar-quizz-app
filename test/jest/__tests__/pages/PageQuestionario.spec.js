/* eslint-disable */

import { mountQuasar } from '../../utils/index'
import Vuex from 'vuex'
import PageQuestionario from 'src/pages/PageQuestionario'
import mutations from '../../../../src/store/questionario/mutations'
import state from '../../../../src/store/questionario/state'
import * as getters from '../../../../src/store/questionario/getters'

let questionOne = {
  id: "CVAKZNtLrxDaaMNWa9v8",
  modulo: "modulos/Jom34TmjNRvbvYEwGo5M",
  nivel: 1,
  referencia: "art. 362",
  respostas: [
    { correta: false, letra: 'a', texto: 'a avó materna' },
    { correta: true, letra: 'b', texto: 'a madastra' },
    { correta: false, letra: 'c', texto: 'o primo' },
    { correta: false, letra: 'd', texto: 'do genro do militar' }
  ],
  texto: "De acordo com o RCONT, são elementos essenciais da continência individual, exceto:"
}

describe('Questionario sem respostas selecionadas', () => {

  let currentQuestion = Object.assign(questionOne, {})

  let options = {
    store: new Vuex.Store({
      modules: {
        questionario: {
          namespaced: true,
          getters: {
            getCurrentQuestion: jest.fn().mockReturnValue(currentQuestion),
            getAnswers: jest.fn().mockReturnValue([]),
            ehPrimeiraQuestao: jest.fn().mockReturnValue(true),
            ehUltimaQuestao: jest.fn().mockReturnValue(false)
          }
        }
      }
    })
  }
  const wrapper = mountQuasar(PageQuestionario, options)

  it('Verifica se alguma resposta foi selecionada', () => {
    expect(wrapper.vm.algumaRespostaSelecionada).toBeFalsy()
  })
  it('Verifica se existia resposta cadastrada', () => {
    expect(wrapper.vm.answer).toBeUndefined()
  })
  it('Verifica se botão de analise esta habilitado', () => {
    expect(wrapper.vm.disableAnalise).toBeFalsy()
  })
  it('Verifica se resposta está analisada', () => {
    expect(wrapper.vm.respostaAnalisada).toBeFalsy()
  })
})

describe('Questionario com resposta selecionada sem estar salva no banco', () => {

  let currentQuestion = Object.assign(questionOne, {})
  currentQuestion.respostas[0].selecionada = true

  let options = {
    store: new Vuex.Store({
      modules: {
        questionario: {
          namespaced: true,
          getters: {
            getCurrentQuestion: jest.fn().mockReturnValue(currentQuestion),
            getAnswers: jest.fn().mockReturnValue([]),
            ehPrimeiraQuestao: jest.fn().mockReturnValue(true),
            ehUltimaQuestao: jest.fn().mockReturnValue(false)
          }
        }
      }
    })
  }
  const wrapper = mountQuasar(PageQuestionario, options)

  it('Verifica se alguma resposta foi selecionada', () => {
    expect(wrapper.vm.algumaRespostaSelecionada).toBeTruthy()
  })
  it('Verifica se botão de analise esta habilitado', () => {
    expect(wrapper.vm.disableAnalise).toBeFalsy()
  })
  it('Verifica se resposta está analisada', () => {
    expect(wrapper.vm.respostaAnalisada).toBeFalsy()
  })
})

describe('Questionario com resposta selecionada já salva no banco', () => {

  let currentQuestion = Object.assign(questionOne, {})
  currentQuestion.respostas[0].selecionada = true

  let answer = {
    correta: false,
    idQuestao: 'CVAKZNtLrxDaaMNWa9v8',
    idUsuario: 'gOru4vlt5Jzzlw64i30P',
    letra: 'a',
    modulo: 'modulos/Jom34TmjNRvbvYEwGo5M'
  }

  let options = {
    store: new Vuex.Store({
      modules: {
        questionario: {
          namespaced: true,
          getters: {
            getCurrentQuestion: jest.fn().mockReturnValue(currentQuestion),
            getAnswers: jest.fn().mockReturnValue([answer]),
            ehPrimeiraQuestao: jest.fn().mockReturnValue(true),
            ehUltimaQuestao: jest.fn().mockReturnValue(false)
          }
        }
      }
    })
  }
  const wrapper = mountQuasar(PageQuestionario, options, true)

  it('Verifica se alguma resposta foi selecionada', () => {
    expect(wrapper.vm.algumaRespostaSelecionada).toBeTruthy()
  })
  it('Verifica se botão de analise esta habilitado', () => {
    expect(wrapper.vm.disableAnalise).toBeTruthy()
  })
  it('Verifica se resposta está analisada', () => {
    expect(wrapper.vm.respostaAnalisada).toBeTruthy()
  })
})

describe('Questionario com 2 questões, alterando entre questões', () => {

  let questionTwo = Object.assign(questionOne, {})
  questionTwo.texto = 'Segundo o RISAER, o luto é o afastamento total do serviço concedido pelo Comandante, Chefe, Diretor ou Prefeito da OM tão logo tenha conhecimento do óbito,'
  questionTwo.id = 'CVAKZNtLrxDaaMNWa9v8'
  state.questions = [questionOne, questionTwo]

  let options = {
    store: new Vuex.Store({
      modules: {
        questionario: {
          namespaced: true,
          state,
          mutations,
          getters
        }
      }
    })
  }
  const wrapper = mountQuasar(PageQuestionario, options, true)
  wrapper.vm.next()

  it('Renderiza corretamente', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Verifica resposta analisada segunda questão', () => {
    expect(wrapper.vm.respostaAnalisada).toBeFalsy()
  })
  it('Verifica se é ultima questão', () => {
    expect(wrapper.vm.ehUltimaQuestao).toBeTruthy()
  })
  it('Verifica se não é primeira questão', () => {
    expect(wrapper.vm.ehPrimeiraQuestao).toBeFalsy()
  })

})