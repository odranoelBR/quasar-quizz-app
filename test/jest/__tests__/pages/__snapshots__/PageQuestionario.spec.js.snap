// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questionario com 2 questões, alterando entre questões Renderiza corretamente 1`] = `
<main
  class="q-page"
  id="questionario"
  style="min-height: 768px;"
>
  <hr
    aria-orientation="horizontal"
    class="q-separator q-mb-md q-separator q-separator--horizontal q-separator--dark"
    role="separator"
  />
   
  <div>
    <span
      class="text-white texto-courier"
      style="font-size: 11pt;"
    >
      Segundo o RISAER, o luto é o afastamento total do serviço concedido pelo Comandante, Chefe, Diretor ou Prefeito da OM tão logo tenha conhecimento do óbito,
    </span>
     
    <div
      class="row"
    >
      <span
        class="text-caption text-blue-grey-3"
      >
        
      art. 362
    
      </span>
    </div>
     
    <hr
      aria-orientation="horizontal"
      class="q-separator q-pb-md bg-info q-separator q-separator--horizontal"
      role="separator"
    />
     
    <div
      class="row question-row"
    >
      <div
        class="q-chip row inline no-wrap items-center fit bg-primary text-white q-chip--colored q-chip--selected q-chip--clickable cursor-pointer non-selectable q-hoverable q-chip--square"
        tabindex="0"
      >
        <div
          class="q-focus-helper"
        />
        <i
          aria-hidden="true"
          class="q-chip__icon q-chip__icon--left material-icons q-icon notranslate"
          role="presentation"
        >
          check
        </i>
        <div
          class="q-chip__content col row no-wrap items-center q-anchor--skip"
        >
          <div
            class="q-avatar full-height"
          >
            <div
              class="q-avatar__content row flex-center overflow-hidden"
            >
              <i
                aria-hidden="true"
                class="mdi mdi-alpha-a q-icon notranslate"
                role="presentation"
              >
                 
              </i>
              
        )
      
            </div>
          </div>
           
          <span
            class="texto-courier"
          >
            a avó materna
          </span>
        </div>
      </div>
    </div>
    <div
      class="row question-row"
    >
      <div
        class="q-chip row inline no-wrap items-center fit text-primary q-chip--colored q-chip--outline q-chip--square"
      >
        <div
          class="q-chip__content col row no-wrap items-center q-anchor--skip"
        >
          <div
            class="q-avatar full-height"
          >
            <div
              class="q-avatar__content row flex-center overflow-hidden"
            >
              <i
                aria-hidden="true"
                class="mdi mdi-alpha-b q-icon notranslate"
                role="presentation"
              >
                 
              </i>
              
        )
      
            </div>
          </div>
           
          <span
            class="texto-courier"
          >
            a madastra
          </span>
        </div>
      </div>
    </div>
    <div
      class="row question-row"
    >
      <div
        class="q-chip row inline no-wrap items-center fit text-primary q-chip--colored q-chip--outline q-chip--square"
      >
        <div
          class="q-chip__content col row no-wrap items-center q-anchor--skip"
        >
          <div
            class="q-avatar full-height"
          >
            <div
              class="q-avatar__content row flex-center overflow-hidden"
            >
              <i
                aria-hidden="true"
                class="mdi mdi-alpha-c q-icon notranslate"
                role="presentation"
              >
                 
              </i>
              
        )
      
            </div>
          </div>
           
          <span
            class="texto-courier"
          >
            o primo
          </span>
        </div>
      </div>
    </div>
    <div
      class="row question-row"
    >
      <div
        class="q-chip row inline no-wrap items-center fit text-primary q-chip--colored q-chip--outline q-chip--square"
      >
        <div
          class="q-chip__content col row no-wrap items-center q-anchor--skip"
        >
          <div
            class="q-avatar full-height"
          >
            <div
              class="q-avatar__content row flex-center overflow-hidden"
            >
              <i
                aria-hidden="true"
                class="mdi mdi-alpha-d q-icon notranslate"
                role="presentation"
              >
                 
              </i>
              
        )
      
            </div>
          </div>
           
          <span
            class="texto-courier"
          >
            do genro do militar
          </span>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="row q-pa-xs q-gutter-sm justify-between text-right"
  >
    <transition-stub
      appear="true"
      duration="[object Object]"
      enteractiveclass="animated fadeIn"
      leaveactiveclass="animated fadeOut"
      mode="out-in"
    >
      <div
        class="col-auto"
      >
        <button
          class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-positive text-white q-btn--actionable q-focusable q-hoverable q-btn--wrap"
          role="button"
          tabindex="0"
          type="button"
        >
          <span
            class="q-focus-helper"
          />
          <span
            class="q-btn__wrapper col row q-anchor--skip"
          >
            <span
              class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
            >
              
          Analisar
        
            </span>
          </span>
        </button>
      </div>
    </transition-stub>
     
    <div
      class="col"
    >
      <button
        class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-accent text-white q-btn--actionable q-focusable q-hoverable q-btn--wrap"
        role="button"
        style="display: none;"
        tabindex="0"
        type="button"
      >
        <span
          class="q-focus-helper"
        />
        <span
          class="q-btn__wrapper col row q-anchor--skip"
        >
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="material-icons q-icon notranslate"
              role="presentation"
            >
              arrow_back
            </i>
          </span>
        </span>
      </button>
    </div>
     
    <div
      class="col-auto"
    >
      <button
        class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-accent text-white q-btn--actionable q-focusable q-hoverable q-btn--wrap"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="q-focus-helper"
        />
        <span
          class="q-btn__wrapper col row q-anchor--skip"
        >
          <span
            class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
          >
            <i
              aria-hidden="true"
              class="material-icons q-icon notranslate"
              role="presentation"
            >
              arrow_forward
            </i>
          </span>
        </span>
      </button>
    </div>
  </div>
</main>
`;
